src/common/providers/identity.ts

```diff
@@ -109,3 +375,598 @@ export function userRecordConstructor(wireData: Object): UserRecord {
+  const publicKey = getPublicKey(header, publicKeys);
+  const decoded = jwt.verify(token, publicKey, {
+    algorithms: [this.algorithm],
+  }) as DecodedJwt;
+  decoded.uid = decoded.sub;
+  checkDecodedToken(decoded, projectId);
+
+  if (decoded.event_type !== eventType) {
+    throw new HttpsError(
+      'invalid-argument',
+      `Expected "${eventType}" but received "${decoded.event_type}".`
+    );
+  }
+  return decoded;
+}
+
+/**
+ * @internal
+ * Helper function to parse the decoded metadata object into a UserMetaData object
+ */
+export function parseMetadata(metadata: DecodedJwtMetadata): UserMetadata {
+  const creationTime = metadata?.creation_time
+    ? new Date((metadata.creation_time as number) * 1000).toUTCString()
+    : null;
+  const lastSignInTime = metadata?.last_sign_in_time
+    ? new Date((metadata.last_sign_in_time as number) * 1000).toUTCString()
+    : null;
+  return new UserRecordMetadata(creationTime, lastSignInTime);
+}
+
+/**
+ * @internal
+ * Helper function to parse the decoded user info array into a UserInfo array
+ */
+export function parseProviderData(
+  providerData: DecodedJwtUserInfo[]
+): UserInfo[] {
+  const providers: UserInfo[] = [];
+  for (const provider of providerData) {
+    providers.push(
+      new UserRecordInfo(
+        provider.uid,
+        provider.display_name,
+        provider.email,
+        provider.photo_url,
+        provider.provider_id,
+        provider.phone_number
+      )
+    );
+  }
+  return providers;
+}
+
+/**
+ * @internal
+ * Helper function to parse the date into a UTC string
+ */
+export function parseDate(tokensValidAfterTime?: number): string | null {
+  if (!tokensValidAfterTime) {
+    return null;
+  }
+  tokensValidAfterTime = tokensValidAfterTime * 1000;
+  try {
+    const date = new Date(tokensValidAfterTime);
+    if (!isNaN(date.getTime())) {
+      return date.toUTCString();
+    }
+  } catch {
+    return null;
+  }
+  return null;
+}
+
+/**
+ * @internal
+ * Helper function to parse the decoded enrolled factors into a valid MultiFactorSettings
+ */
+export function parseMultiFactor(
+  multiFactor?: DecodedJwtEnrolledFactors
+): MultiFactorSettings {
+  if (!multiFactor) {
+    return null;
+  }
+  const parsedEnrolledFactors: MultiFactorInfo[] = [];
+  for (const factor of multiFactor.enrolled_factors || []) {
+    if (!factor.uid) {
+      throw new HttpsError(
+        'internal',
+        'INTERNAL ASSERT FAILED: Invalid multi-factor info response'
+      );
+    }
+    const enrollmentTime = factor.enrollment_time
+      ? new Date(factor.enrollment_time).toUTCString()
+      : null;
+    if (factor.phone_number) {
+      parsedEnrolledFactors.push(
+        new UserRecordPhoneMultiFactorInfo(
+          factor.uid,
+          factor.factor_id || 'phone',
+          factor.phone_number,
+          factor.display_name,
+          enrollmentTime
+        ) as PhoneMultiFactorInfo
```

Comment:
All these casts seem like a code smell. Why isn't a UserRecordPhoneMultiFactorInfo already a PhoneMultiFactorInfo?

Reply:
Good point, I end up refactoring all the classes to interfaces and just have `phoneNumber` as a optional property